Mathlib/Data/Set/Pairwise.lean

```diff
@@ -0,0 +1,495 @@
+
+theorem PairwiseDisjoint.eq_of_le (hs : s.PairwiseDisjoint f) {i j : ι} (hi : i ∈ s) (hj : j ∈ s)
+    (hf : f i ≠ ⊥) (hij : f i ≤ f j) : i = j :=
+  (hs.elim' hi hj) fun h => hf <| (inf_of_le_left hij).symm.trans h
+#align set.pairwise_disjoint.eq_of_le Set.PairwiseDisjoint.eq_of_le
+
+end SemilatticeInfBot
+
+section CompleteLattice
+
+variable [CompleteLattice α]
+
+/-- Bind operation for `Set.PairwiseDisjoint`. If you want to only consider finsets of indices, you
+can use `Set.PairwiseDisjoint.bunionᵢ_finset`. -/
+theorem PairwiseDisjoint.bunionᵢ {s : Set ι'} {g : ι' → Set ι} {f : ι → α}
+    (hs : s.PairwiseDisjoint fun i' : ι' => ⨆ i ∈ g i', f i)
+    (hg : ∀ i ∈ s, (g i).PairwiseDisjoint f) : (⋃ i ∈ s, g i).PairwiseDisjoint f := by
+  rintro a ha b hb hab
+  simp_rw [Set.mem_unionᵢ] at ha hb
+  obtain ⟨c, hc, ha⟩ := ha
+  obtain ⟨d, hd, hb⟩ := hb
+  obtain hcd | hcd := eq_or_ne (g c) (g d)
+  · exact hg d hd (hcd.subst ha) hb hab
+  · refine (hs hc hd <| ne_of_apply_ne _ hcd).mono ?_ ?_
+    · let f' := fun (i : ι) (_ : i ∈ g c) => f i
+      change f' a ha ≤ ⨆ (i : ι) (H : i ∈ g c), f' i H
+      exact le_supᵢ₂ a ha
+    · let f'' := fun (i : ι) (_ : i ∈ g d) => f i
+      change f'' b hb ≤ ⨆ (i : ι) (H : i ∈ g d), f'' i H
+      exact le_supᵢ₂ b hb
+
+#align set.pairwise_disjoint.bUnion Set.PairwiseDisjoint.bunionᵢ
+
+end CompleteLattice
+
+/-! ### Pairwise disjoint set of sets -/
+
+
+theorem pairwiseDisjoint_range_singleton :
+    (Set.range (singleton : ι → Set ι)).PairwiseDisjoint id := by
+  rintro _ ⟨a, rfl⟩ _ ⟨b, rfl⟩ h
+  exact disjoint_singleton.2 (ne_of_apply_ne _ h)
+#align set.pairwise_disjoint_range_singleton Set.pairwiseDisjoint_range_singleton
+
+theorem pairwiseDisjoint_fiber (f : ι → α) (s : Set α) : s.PairwiseDisjoint fun a => f ⁻¹' {a} :=
+  fun _a _ _b _ h => disjoint_iff_inf_le.mpr fun _i ⟨hia, hib⟩ => h <| (Eq.symm hia).trans hib
+#align set.pairwise_disjoint_fiber Set.pairwiseDisjoint_fiber
+
+-- classical
+theorem PairwiseDisjoint.elim_set {s : Set ι} {f : ι → Set α} (hs : s.PairwiseDisjoint f) {i j : ι}
+    (hi : i ∈ s) (hj : j ∈ s) (a : α) (hai : a ∈ f i) (haj : a ∈ f j) : i = j :=
+  hs.elim hi hj <| not_disjoint_iff.2 ⟨a, hai, haj⟩
+#align set.pairwise_disjoint.elim_set Set.PairwiseDisjoint.elim_set
+
+theorem bunionᵢ_diff_bunionᵢ_eq {s t : Set ι} {f : ι → Set α} (h : (s ∪ t).PairwiseDisjoint f) :
+    ((⋃ i ∈ s, f i) \ ⋃ i ∈ t, f i) = ⋃ i ∈ s \ t, f i := by
+  refine'
+    (bunionᵢ_diff_bunionᵢ_subset f s t).antisymm
+      (unionᵢ₂_subset fun i hi a ha => (mem_diff _).2 ⟨mem_bunionᵢ hi.1 ha, _⟩)
+  rw [mem_unionᵢ₂]; rintro ⟨j, hj, haj⟩
+  exact (h (Or.inl hi.1) (Or.inr hj) (ne_of_mem_of_not_mem hj hi.2).symm).le_bot ⟨ha, haj⟩
+#align set.bUnion_diff_bUnion_eq Set.bunionᵢ_diff_bunionᵢ_eq
+
+/-- Equivalence between a disjoint bounded union and a dependent sum. -/
+noncomputable def bUnionᵢEqSigmaOfDisjoint {s : Set ι} {f : ι → Set α} (h : s.PairwiseDisjoint f) :
```

Comment:
I wasn't sure how the capitalization rules worked with things like `bUnion`, so I went with this.